ipa-migrate - fix migration issues with entries using ipaUniqueId in the RDN

mreynolds389 opened the pull request:
We need to handle these entries differently and specify what attribute and search base to use to find the entry on the local server.  Most entries can use the "cn" attribute but for selinux usermaps we need to search using the ipaOwner attribute which is a DN, and in turn requires additional handling/converting in order to properly check if the usermap exists or not.

Fixes: https://pagure.io/freeipa/issue/9640

mreynolds389 commented on ipaserver/install/ipa_migrate_constants.py:
```diff
@@ -777,6 +783,11 @@
         'subtree': ',cn=subids,cn=accounts,$SUFFIX',
         'label': 'Sub IDs',
         'mode': 'all',  # TODO Maybe production only?
```
@rcritten - I suspect "subids" should be production only, can you confirm?

rcritten replied:
Yes, that seems like a good choice.

mreynolds389 replied:
> Yes, that seems like a good choice.

Ok I'll make this change then...

rcritten commented:
I primarily tested hbac rules.
1.  The original UUID remains which is fine
2. Marking a rule as disabled is migrated

I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.

mreynolds389 commented:
> I primarily tested hbac rules.
> 
> 1. The original UUID remains which is fine
> 2. Marking a rule as disabled is migrated
> 
> I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.

Hmm, yeah perhaps it's not comparing them.  I'll look into it today...

mreynolds389 commented:
> > I primarily tested hbac rules.
> > 
> > 1. The original UUID remains which is fine
> > 2. Marking a rule as disabled is migrated
> > 
> > I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.
> 
> Hmm, yeah perhaps it's not comparing them. I'll look into it today...

Ok it works for me:  usercat was properly migrated.  

```
Before on local server:

# ipa hbacrule-show markRule
  Rule name: markRule
  Enabled: True

On remove server:

# ipa hbacrule-mod --usercat=all
Rule name: markRule
-----------------------------
Modified HBAC rule "markRule"
-----------------------------
  Rule name: markRule
  User category: all
  Enabled: True



After ipa-mgrate on the local server:

# ipa hbacrule-show markRule
  Rule name: markRule
  User category: all
  Enabled: True
```

mreynolds389 commented:
> > > I primarily tested hbac rules.
> > > 
> > > 1. The original UUID remains which is fine
> > > 2. Marking a rule as disabled is migrated
> > > 
> > > I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.
> > 
> > 
> > Hmm, yeah perhaps it's not comparing them. I'll look into it today...
> 
> Ok it works for me: usercat was properly migrated.
> 
> ```
> Before on local server:
> 
> # ipa hbacrule-show markRule
>   Rule name: markRule
>   Enabled: True
> 
> On remove server:
> 
> # ipa hbacrule-mod --usercat=all
> Rule name: markRule
> -----------------------------
> Modified HBAC rule "markRule"
> -----------------------------
>   Rule name: markRule
>   User category: all
>   Enabled: True
> 
> 
> 
> After ipa-mgrate on the local server:
> 
> # ipa hbacrule-show markRule
>   Rule name: markRule
>   User category: all
>   Enabled: True
> ```

Ok, sorry I misunderstood.  You were trying to set it to "allow_all".  When I try it says that's not a valid option:

```
# ipa hbacrule-mod --usercat=allow_all
Rule name: markRule
ipa: ERROR: invalid 'usercat': must be 'all'
```

Now I realize you were trying to update the "allow_all" rule. 

```
# ipa hbacrule-mod --usercat=  allow_all
------------------------------
Modified HBAC rule "allow_all"
------------------------------
  Rule name: allow_all
  Host category: all
  Service category: all
  Description: Allow all users to access any host from any host
  Enabled: True

```

 By bad, continuing investigation....

mreynolds389 commented:
> I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.

@rcritten Ah yes, there is a design "flaw", but fixing it could be risky.  Basically we only add what is missing to the local entry - we don't remove anything from the local entry.  We should definitely fix it, but this is a change in behavior and will require more testing...

mreynolds389 commented:
> > I did discover an issue. I ran `ipa hbacrule-mod --usercat= allow_all` and the migrated entry still had usercat=all.
> 
> @rcritten Ah yes, there is a design "flaw", but fixing it could be risky. Basically we only add what is missing to the local entry - we don't remove anything from the local entry. We should definitely fix it, but this is a change in behavior and will require more testing...


I have an initial fix (not pushed), but it is removing some attributes that are "sort of" unexpected:

```
2024-07-31T16:24:47Z DEBUG Entry is different and will be updated: 'uid=admin,cn=users,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'passwordGraceUserTime' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'uid=sudo,cn=sysaccounts,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'passwordExpirationTime' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'uid=sudo,cn=sysaccounts,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsIdleTimeout' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'ipaUniqueID=d7d9de00-4e8a-11ef-9920-3cd92bf69aa8,cn=hbac,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCategory' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'krbprincipalname=ldap/hpe-bl420cgen8-01.khw.eng.rdu2.dc.redhat.com@KHW.ENG.RDU2.DC.REDHAT.COM,cn=services,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCertificate' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'krbprincipalname=HTTP/hpe-bl420cgen8-01.khw.eng.rdu2.dc.redhat.com@KHW.ENG.RDU2.DC.REDHAT.COM,cn=services,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCertificate' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'cn=anonymous-limits,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsSizeLimit' is being removed
2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'cn=anonymous-limits,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsLookThroughLimit' is being removed
```

rcritten commented:
> I have an initial fix (not pushed), but it is removing some attributes that are "sort of" unexpected:
> 
> ```
> 2024-07-31T16:24:47Z DEBUG Entry is different and will be updated: 'uid=admin,cn=users,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'passwordGraceUserTime' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'uid=sudo,cn=sysaccounts,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'passwordExpirationTime' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'uid=sudo,cn=sysaccounts,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsIdleTimeout' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'ipaUniqueID=d7d9de00-4e8a-11ef-9920-3cd92bf69aa8,cn=hbac,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCategory' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'krbprincipalname=ldap/hpe-bl420cgen8-01.khw.eng.rdu2.dc.redhat.com@KHW.ENG.RDU2.DC.REDHAT.COM,cn=services,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCertificate' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'krbprincipalname=HTTP/hpe-bl420cgen8-01.khw.eng.rdu2.dc.redhat.com@KHW.ENG.RDU2.DC.REDHAT.COM,cn=services,cn=accounts,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'userCertificate' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'cn=anonymous-limits,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsSizeLimit' is being removed
> 2024-07-31T16:24:48Z DEBUG Entry is different and will be updated: 'cn=anonymous-limits,cn=etc,dc=khw,dc=eng,dc=rdu2,dc=dc,dc=redhat,dc=com' attribute 'nsLookThroughLimit' is being removed
> ```

Curious indeed. Does this suggest that the setting is in the "local" installation and not the remote one, or the other way around?

I suspect that for userCertificate you've already filtered it out as having been issued by the remote CA and now it is noticing that only the local entry has one so it wants to remove that one too.

mreynolds389 commented:
> Curious indeed. Does this suggest that the setting is in the "local" installation and not the remote one, or the other way around?

The local server is the one we are updating - the new installation.  The remote server is where we pull in the data we want to migrate.  In this case the local server has attributes that are not present on the remote server/entry.  So in order to make it match we remove these attributes from the new deployment.  

> 
> I suspect that for userCertificate you've already filtered it out as having been issued by the remote CA and now it is noticing that only the local entry has one so it wants to remove that one too.

Yes, the remote entry had been "cleaned" prior to this step.  So at this stage the local entry has a certificate, when the remote does not, so it removes it.  I don't know if there are cases that we would to keep the cert.  

Since both IPA servers are the same I was surprised to see the resource limit attributes being removed (nsSizeLimit, passwordGrace/expirationtime, etc).  I suspect the remote entry is missing the operational attributes so I'll look into that and see if I can clean it up a bit.

rcritten commented:
We need to keep the userCertficate value(s) in the local entry.

rcritten commented:
I added a caacl rule on remote to test with and ran the first migration. It's working fine.
I tweaked a rule on remote and ran the migration again and migration failed:
`Failed to update "krbprincipalname=DNS/replica.example.test@EXAMPLE.TEST,cn=services,cn=accounts,dc=example,dc=test" error: missing attribute "ipaKrbPrincipalAlias" required by object class "ipaKrbPrincipal"`

replica in this case is the "local" IPA server. This service entry is not present on remote.

```
[remote] $ ipa service-find replica.example.test
------------------
0 services matched
------------------
----------------------------
Number of entries returned 0
----------------------------
```

rcritten commented:
When migrating a sudorule the order attribute is additive. So if the rule test exists both on local and remote with different values the result contains both in order, e.g. 1, 9. I think because sudoOrder is not defined as single:

`attributetypes: ( 1.3.6.1.4.1.15953.9.1.10 NAME 'sudoOrder' DESC 'an integer to order the sudoRole entries' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 X-ORIGIN 'SUDO' )`

mreynolds389 commented:
> When migrating a sudorule the order attribute is additive. So if the rule test exists both on local and remote with different values the result contains both in order, e.g. 1, 9. I think because sudoOrder is not defined as single:
> 
> `attributetypes: ( 1.3.6.1.4.1.15953.9.1.10 NAME 'sudoOrder' DESC 'an integer to order the sudoRole entries' EQUALITY integerMatch ORDERING integerOrderingMatch SYNTAX 1.3.6.1.4.1.1466.115.121.1.27 X-ORIGIN 'SUDO' )`

Yeah we check the schema to determine how to perform the update.  So unfortunately this attribute will need special processing.  Are there other attributes like this? :-(

mreynolds389 commented:
> I added a caacl rule on remote to test with and ran the first migration. It's working fine. I tweaked a rule on remote and ran the migration again and migration failed: `Failed to update "krbprincipalname=DNS/replica.example.test@EXAMPLE.TEST,cn=services,cn=accounts,dc=example,dc=test" error: missing attribute "ipaKrbPrincipalAlias" required by object class "ipaKrbPrincipal"`

What change did you make?   I can not reproduce this failure.  I added a new caacl rule, ran migration, then I changed it to add "all" to hostcat and ran migration, and then I remove the hostcat and ran migration - it worked fine.  I'm sure the problem is real, but I don't know how to reproduce it .  Also I don't see how a caacl change would impact a service entry.  A

Any chance you still have the migration log?

rcritten commented:
We override/set a few other single-value attributes at https://github.com/freeipa/freeipa/blob/master/ipapython/ipaldap.py#L760

These are almost all, I think, not defined in any schema. The way that sudoOrder is used in IPA should always remain single-valued but I suppose you might be able to extend this list to include it.

rcritten commented:
My test rule is:
```
# ipa caacl-show test
  ACL name: test
  Enabled: True
  Host category: all
```

mreynolds389 commented:
> My test rule is:
> 
> ```
> # ipa caacl-show test
>   ACL name: test
>   Enabled: True
>   Host category: all
> ```

Sorry, I added this test rule on the remote server and the migration works fine with the current patch.  Also those other "single-valued" attributes are for cn=config and that is handled differently.  So AFAICT this PR is ready for review again...

mreynolds389 commented:
@rcritten ok so there are two ways to fix the latest issue with removing an attribute when there's an objectclass that requires it...

[1]  Do not remove the attribute on the local server.  This is the current fix I just pushed
[2]  On the local server remove the attribute AND the objectclass that requires it

Like I said I just merged option [1], but we could do option [2].  I can see cases where both might be useful.

Thoughts?

rcritten commented:
I think your approach is fine and it is working for me. You also addressed the re-run migration issue I was having.

This does raise an interesting question about what we will and won't support when it comes to migration. I imagine that some users who have dev, int, prod systems might want to re-sync their dev and int systems periodically from live. While running migration multiple times seems to work for me, and prefers the existing "remote" values over any "local" value I wonder if some users will chafe at that. I'm not saying we change anything. I wonder if we need to write somewhere that the "remote" side preferred so in case of differences, it wins.
For example, local I have hbacrule test with no options set. On remote I have same rule with --hostcat=all.
Upon migration the local rule is now hostcat=all. This is obviously what we want and intend, I just don't know how clearly we state it, if anywhere.

That is outside the scope of the PR. Just something to chew on.

I'll ack once the tests are complete.

mreynolds389 commented:
> This does raise an interesting question about what we will and won't support when it comes to migration. I imagine that some users who have dev, int, prod systems might want to re-sync their dev and int systems periodically from live. While running migration multiple times seems to work for me, and prefers the existing "remote" values over any "local" value I wonder if some users will chafe at that. I'm not saying we change anything. I wonder if we need to write somewhere that the "remote" side preferred so in case of differences, it wins. For example, local I have hbacrule test with no options set. On remote I have same rule with --hostcat=all. Upon migration the local rule is now hostcat=all. This is obviously what we want and intend, I just don't know how clearly we state it, if anywhere.

I have to double check but I think in the design doc it states that the goal of the migration tool to take what is in the remote server and sync it to the local one.  So the local server should always be expected to be overwritten.  And it would be difficult to change this behavior, or make it switchable.  I think we just need to get this thing out into the wild and see how people use it.  I'm sure they will come across more corner cases, but I really feel like these last set of changes addressed most of the remaining issues.  So I think we are ready to ship it :-)  Well, after QE does their testing day of course...

Oh, but I do have one last question/concern about the migration log.  I'm still undecided about it always appending each migration run.  I find it annoying to go through the log after multiple migrations have been run, and I wish it just overwrote it on each migration attempt.  This also means that log will grow forever and never be rotated/trimmed.  There are pro's and cons to both approaches.  This is not critical, but its been nagging me haha.  Well, it is definitely something that can be discussed later...

flo-renaud commented:
master:

* 7808fc8398b54a9008872c3d5cb13ccde4ec10bc ipa-migrate - fix migration issues with entries using ipaUniqueId in the RDN

flo-renaud commented:
Hi @mreynolds389 
a manual backport is required for ipa-4-12 branch, can you handle it? Thanks

mreynolds389 commented:
> Hi @mreynolds389 a manual backport is required for ipa-4-12 branch, can you handle it? Thanks

ipa-4-12  fba36897744a8fcdd170468b6c6434fb42838450
